File: ios/Bain Luck/Bain Luck/Utilities/DrawPricedWinner.swift
import Foundation

/// What a match surface may print when the winner market has THREE outcomes.
///
/// #5271. Every probability the event page draws for a match came from one
/// pair, and the pair is `home` and `1 − home`. On a two-outcome sport that is
/// exactly right, and the app leans on it hard — `routes/feed.py` derives
/// `current_odds.away_probability` as `round(1.0 - current_home_prob, 6)`, and
/// `renderedDuelPercents` is built on the two summing to one.
///
/// `1 − P(home)` is **"the home team does not win"**. In soccer that is *away
/// win **or** draw*, so the figure printed under the away crest silently
/// absorbs the entire draw probability.
///
/// Photographed on production 2026-09-11 (event `15304603`, Al-Faisaly KSA FC
/// v Al-Ittihad, live 0–2). One screen, one source — `win_probability_sources`
/// held `{"kalshi": {"value": 0.01}}` and nothing else:
///
/// | where | Al-Ittihad | Tie | Al-Faisaly |
/// |---|---|---|---|
/// | hero, top of page | **99%** | — | 1% |
/// | Other Markets card, same page | **93%** | 5% | 1% |
///
/// The reader scrolls two cards and the favourite loses six points. Alex found
/// the same 99% a third time in the goal-margin map's header. Mid-game at 0–2
/// the draw is nearly dead so the error is ~6pp; **pre-match a league draw
/// prices around 20–30%**, and every scheduled soccer fixture the app renders
/// is affected.
///
/// ═══ WHY THIS WITHHOLDS RATHER THAN CORRECTS ═══
///
/// The obvious fix — print the away team's real price — has nowhere to read one
/// from. Measured 2026-09-11 before this shipped:
///
///  * the event payload carries no draw and no independent away price: the away
///    field IS the complement;
///  * `win_prob_snapshots.draw_probability` exists, is served on the history
///    payload this view already fetches, and is **NULL on all 3,084,750 rows** —
///    no writer has ever set it;
///  * the three-way groups on `game-markets` are not fit to print (#5328): of
///    29 sampled soccer events only 13 served a tie group at all, and of four
///    read in full one live match was "priced" at 99% draw and two summed to
///    0.87 and 1.04.
///
/// So the honest move is the one this codebase already makes wherever it cannot
/// state a number — `SportVocab.totalRange`'s `nil`, `EventOutcome.undecided`,
/// `formatProbabilityOrDash` — *a number in the wrong unit is worse than an
/// absent one, because it looks sourced.* The away slot is withheld.
///
/// ═══ WHAT THIS DOES NOT FIX, AND MUST NOT BE READ AS FIXING ═══
///
/// The home price this keeps is the blend, and on soccer the blend is itself
/// draw-dropped whenever the books are in it (#1011). Measured the same day
/// over recent soccer events carrying both sources, `betting` sat **10–21pp
/// above** the Kalshi three-way home price on 8 of 8 sampled — Flamengo 0.856
/// against 0.650, Atletico Mineiro 0.849 against 0.685 — which is a `betting`
/// source normalised two ways with the draw discarded, at weight 3.0 against
/// Kalshi's 0.8.
///
/// That is a DATA defect and it is #1011's. This type fixes the RENDER defect
/// only: a client inventing an away price out of a complement. The two are
/// independent, and the issue's acceptance test — hero and match-winner card
/// agreeing on one screen — is only reachable once #1011 lands too.
enum DrawPricedWinner {

    /// Whether this sport's match-winner market prices a draw as a third
    /// outcome. Declared per sport in `SportVocab`, never inferred from a key.
    static func sportPricesADraw(_ sportKey: String?) -> Bool {
        SportVocab.forSport(sportKey).winnerMarketPricesADraw
    }

    /// The two win probabilities a two-slot surface may print, or `nil` where
    /// it may print no probabilities at all.
    ///
    /// A `nil` `away` inside a non-nil result is the whole point: it means *the
    /// home number stands and the away slot is withheld*, which is a different
    /// answer from "this surface has no numbers" and the callers draw them
    /// differently.
    ///
    /// - Parameter away: the away figure the caller would otherwise have
    ///   printed — the served `current_odds.away_probability`, or a locally
    ///   derived `1 − home`. It is IGNORED on a draw-priced sport; the
    ///   parameter stays so that two-way callers keep passing the served value
    ///   rather than re-deriving one, which is the trap `duelPercents`
    ///   documents.
    /// - Returns: `nil` when there is no home price, and — preserving today's
    ///   behaviour exactly — when a two-way sport holds no away price either.
    static func printablePair(
        away: Double?,
        home: Double?,
        sport sportKey: String?
    ) -> (away: Double?, home: Double)? {
        guard let home else { return nil }
        if sportPricesADraw(sportKey) { return (away: nil, home: home) }
        guard let away else { return nil }
        return (away: away, home: home)
    }

    /// The ONE side a single-number headline may name, and its probability.
    ///
    /// `MarketMapView`'s map header names the favourite — the side above 50% —
    /// and on a draw-priced sport that choice reaches for the complement every
    /// time the home team is the underdog, which is how Alex found
    /// `Al-Ittihad 99%` sitting above a match its own card priced at 93%.
    ///
    /// A draw-priced sport therefore always names HOME, favourite or not. A
    /// headline that names the underdog is honest; one that names the wrong
    /// number is not.
    static func headlineSide(
        away: Double?,
        home: Double?,
        sport sportKey: String?
    ) -> (isHome: Bool, probability: Double)? {
        guard let pair = printablePair(away: away, home: home, sport: sportKey) else { return nil }
        guard let away = pair.away else { return (isHome: true, probability: pair.home) }
        let homeIsFavoured = pair.home > 0.5
        return (isHome: homeIsFavoured, probability: homeIsFavoured ? pair.home : away)
    }

    /// Who the hero's "since open" caption names, and the move it states.
    ///
    /// #1830 — the caption used to be a bare unlabelled `-27%` computed on
    /// HOME, printed under an away-first `87 – 13` hero, so Alex read the Red
    /// Sox as FALLING 27 while they had gone 60 → 87. The fix was to name the
    /// team and report the RISER, always as a gain: because `away == 1 − home`
    /// exactly, "home fell 27" and "away rose 27" are the same fact.
    ///
    /// **That equivalence is exactly what a draw breaks.** Home shedding five
    /// points does not hand them to the away side; the draw can take any part
    /// of them. So a draw-priced sport names HOME and signs the number, which
    /// is still no bare figure — #1830's defect was the missing subject, not
    /// the minus sign.
    ///
    /// - Parameter homeDelta: home now minus home at open, as a probability.
    /// - Returns: which side to name, and the whole points to print WITH the
    ///   sign the caption should show (never negative on a two-way sport).
    static func trendSubject(
        homeDelta: Double,
        sport sportKey: String?
    ) -> (isHome: Bool, signedPoints: Int) {
        let points = Int((abs(homeDelta) * 100).rounded())
        if sportPricesADraw(sportKey) {
            return (isHome: true, signedPoints: homeDelta < 0 ? -points : points)
        }
        return (isHome: homeDelta > 0, signedPoints: points)
    }
}

File: frontend/lib/drawPricedWinner.ts
import { sportVocab } from "@/lib/marketMapUtils";

/**
 * WHAT A MATCH SURFACE MAY PRINT WHEN THE WINNER MARKET HAS THREE OUTCOMES.
 *
 * #6238 — the WEB twin of native's #5271 (`ios/.../Utilities/DrawPricedWinner.swift`,
 * shipped 2026-09-11). The guard existed in one client and the other client was
 * unguarded: before this file, the only path outside `ios/` that named the flag
 * was `__tests__/ios/aDrawIsNotTheAwayTeam5271.test.ts`, a jest test that scans
 * the *Swift source*. So nothing went red while the web printed the complement.
 *
 * ═══ THE DEFECT ═══
 *
 * Every probability this site draws for a match comes from one pair, and the
 * pair is `home` and `1 − home`. `routes/feed.py` derives
 * `current_odds.away_probability` as `round(1.0 - current_home_prob, 6)`, and
 * `renderedDuelPercents` is built on the two summing to one. On a two-outcome
 * sport that is exactly right.
 *
 * `1 − P(home)` is **"the home team does not win"**. In soccer that is *away win
 * **or** draw*, so the figure printed under the away crest silently absorbs the
 * entire draw probability.
 *
 * Photographed on production 2026-09-14 by authority/322 under D48 —
 * `/events/15301234`, León v Atlético San Luis, Liga MX, 2h pre-match, 390px:
 *
 * | where | León | draw | San Luis |
 * |---|---|---|---|
 * | hero, top of page | **68%** | — | **32%** |
 * | the same page's correct-score card | — | 0-0 at 22%, 1-1 at 10% | — |
 *
 * The page refutes itself one card down. The books behind it were nowhere near a
 * two-way split: draftkings −115/+285 implies 53.5%/26.0%, **summing to 79.5%**;
 * fanduel 80.9%; lowvig 77.3%. We divide by that sum to strip vig — correct for a
 * two-way market, wrong here, because the missing ~20 points are mostly the draw
 * and not vig. León was overstated by ~14pp, San Luis by ~6pp.
 *
 * Worse once settled: `/events/15305024` (Daejeon Citizen v Pohang Steelers)
 * finished **2–2**, and its chart readout said `Citizen 63% — Steelers 37%` off
 * terminal odds of **+600 / +950** — both long shots, because in the dying
 * minutes of a level game the draw was the overwhelming favourite and then the
 * actual result.
 *
 * ═══ WHY THIS WITHHOLDS RATHER THAN CORRECTS ═══
 *
 * The obvious fix — print the away team's real price — has nowhere to read one
 * from. Measured for #5271 on 2026-09-11 and re-confirmed on this site's own
 * payloads on 2026-09-14:
 *
 *  * the event payload carries no draw and no independent away price: the away
 *    field IS the complement (`0.317 === 1 − 0.683` on the specimen, exactly);
 *  * `win_prob_snapshots.draw_probability` exists and is **NULL on all 3,084,750
 *    rows** — no writer has ever set it;
 *  * the three-way groups on `game-markets` are not fit to print (#5328).
 *
 * So this takes the move the codebase already makes wherever it cannot state a
 * number — `UNSCORED_IN_POINTS`'s empty unit, `NO_READING`, `formatProbabilityOrDash`:
 * **a number in the wrong unit is worse than an absent one, because it looks
 * sourced.** The away slot is withheld and prints the em-dash the hero already
 * draws for a side with no reading.
 *
 * ═══ WHAT THIS DOES NOT FIX, AND MUST NOT BE READ AS FIXING ═══
 *
 * The home price it KEEPS is the blend, and on soccer the blend is itself
 * draw-dropped whenever the books are in it (**#1011**) — that is why the
 * specimen's retained number is 68% and not the ~53% the raw moneylines imply.
 * This file fixes the RENDER defect only: a client inventing an away price out
 * of a complement. The two are independent, and #6238's acceptance test — the
 * hero and the three-way card agreeing on one screen — is only reachable once
 * #1011 lands too. Nobody should read this landing as that being delivered.
 */

/**
 * Whether this sport's match-winner market prices a draw as a third outcome.
 *
 * Declared per sport in `sportVocab`, **never inferred from a key here**. A
 * `sportKey.startsWith("soccer")` at a call site is the same defect one layer
 * down: it is a spelling test standing in for a semantic one, and it puts the
 * declaration wherever the last person happened to need it. `sportVocab` already
 * matches `soccer_mexico_ligamx` and `soccer_korea_kleague1` through its
 * `["soccer", "mls", "epl", "uefa", "fifa"]` row, which is why both production
 * specimens are covered without anyone listing a league.
 *
 * An undeclared sport answers `false` and keeps its two-sided reading, so
 * widening the rule is one `true` in the vocab and a test.
 */
export function sportPricesADraw(sportKey: string | null | undefined): boolean {
  return sportVocab(sportKey || undefined).winnerMarketPricesADraw;
}

/**
 * The away figure a two-slot surface may print — the served one, or `null` where
 * it may print none.
 *
 * ═══ WHY THIS TAKES THE AWAY VALUE AND RETURNS ONE, RATHER THAN MIRRORING
 *     NATIVE'S `printablePair` ═══
 *
 * Native's `printablePair` returns the whole pair and answers `nil` for the
 * *whole thing* when either side is missing on a two-way sport. Porting that
 * shape literally would change this site's two-way behaviour: the web hero today
 * prints `68% – —%` when the away price alone is absent, and `printablePair`'s
 * contract would withhold both numbers instead. That is a regression on every
 * sport this issue is not about.
 *
 * So the web takes the narrower question, which is the only one #6238 asks: *may
 * this surface print the away number?* Native does exactly this at the two call
 * sites where the difference bites — `DiscoverEventCard.swift:157` reads
 * `sportPricesADraw` directly for this reason, and `EventCardView.swift:597`
 * documents it as "the one place in the family where the difference bites". This
 * is that same choice, made once for the web rather than at each call site.
 *
 * The `away` parameter is passed through rather than re-derived so that callers
 * keep handing over the SERVED value; a surface that re-derives `1 − home`
 * locally is the trap `renderedDuelPercents` documents.
 */
export function printableAway<T>(away: T | null, sportKey: string | null | undefined): T | null {
  return sportPricesADraw(sportKey) ? null : away;
}
